File: backend/app/core/enums.py
"""Domain enumerations shared by models, schemas and services."""

from __future__ import annotations

from enum import Enum


class StrEnum(str, Enum):
    def __str__(self) -> str:  # pragma: no cover - trivial
        return self.value


class UserRole(StrEnum):
    CITIZEN = "CITIZEN"
    ADMIN = "ADMIN"
    REPAIR_TEAM = "REPAIR_TEAM"


class DamageType(StrEnum):
    POTHOLE = "POTHOLE"
    CRACKED_ROAD = "CRACKED_ROAD"
    FLOODING = "FLOODING"
    DAMAGED_SIDEWALK = "DAMAGED_SIDEWALK"
    BROKEN_STREETLIGHT = "BROKEN_STREETLIGHT"
    OTHER = "OTHER"
    UNKNOWN = "UNKNOWN"


class ComplaintStatus(StrEnum):
    PENDING = "PENDING"
    AI_ANALYZED = "AI_ANALYZED"
    PRIORITIZED = "PRIORITIZED"
    ASSIGNED = "ASSIGNED"
    IN_PROGRESS = "IN_PROGRESS"
    RESOLVED = "RESOLVED"
    REJECTED = "REJECTED"
    DUPLICATE = "DUPLICATE"


#: Allowed status transitions.  Enforced by ``ComplaintService`` so the
#: lifecycle cannot be short-circuited through the API.
STATUS_TRANSITIONS: dict[ComplaintStatus, set[ComplaintStatus]] = {
    ComplaintStatus.PENDING: {
        ComplaintStatus.AI_ANALYZED,
        ComplaintStatus.PRIORITIZED,
        ComplaintStatus.REJECTED,
        ComplaintStatus.DUPLICATE,
    },
    ComplaintStatus.AI_ANALYZED: {
        ComplaintStatus.PRIORITIZED,
        ComplaintStatus.REJECTED,
        ComplaintStatus.DUPLICATE,
    },
    ComplaintStatus.PRIORITIZED: {
        ComplaintStatus.ASSIGNED,
        ComplaintStatus.REJECTED,
        ComplaintStatus.DUPLICATE,
    },
    ComplaintStatus.ASSIGNED: {
        ComplaintStatus.IN_PROGRESS,
        ComplaintStatus.PRIORITIZED,
        ComplaintStatus.REJECTED,
        ComplaintStatus.DUPLICATE,
    },
    ComplaintStatus.IN_PROGRESS: {
        ComplaintStatus.RESOLVED,
        ComplaintStatus.ASSIGNED,
        ComplaintStatus.REJECTED,
    },
    ComplaintStatus.RESOLVED: {ComplaintStatus.IN_PROGRESS},
    ComplaintStatus.REJECTED: {ComplaintStatus.PENDING, ComplaintStatus.PRIORITIZED},
    ComplaintStatus.DUPLICATE: {ComplaintStatus.PENDING, ComplaintStatus.PRIORITIZED},
}

#: Statuses that mean "no longer needs field work".
CLOSED_STATUSES = {ComplaintStatus.RESOLVED, ComplaintStatus.REJECTED, ComplaintStatus.DUPLICATE}


class PriorityLevel(StrEnum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"


class TrafficLevel(StrEnum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    VERY_HIGH = "VERY_HIGH"


class PlaceType(StrEnum):
    HOSPITAL = "HOSPITAL"
    SCHOOL = "SCHOOL"
    BUS_STOP = "BUS_STOP"
    MAJOR_INTERSECTION = "MAJOR_INTERSECTION"
    EMERGENCY_SERVICE = "EMERGENCY_SERVICE"


class AssignmentStatus(StrEnum):
    ASSIGNED = "ASSIGNED"
    IN_PROGRESS = "IN_PROGRESS"
    COMPLETED = "COMPLETED"
    VERIFIED = "VERIFIED"
    CANCELLED = "CANCELLED"


class DuplicateStatus(StrEnum):
    SUGGESTED = "SUGGESTED"
    CONFIRMED = "CONFIRMED"
    REJECTED = "REJECTED"


class NotificationChannel(StrEnum):
    IN_APP = "IN_APP"
    EMAIL = "EMAIL"
    SMS = "SMS"
    PUSH = "PUSH"


class ImageKind(StrEnum):
    REPORT = "REPORT"
    REPAIR_EVIDENCE = "REPAIR_EVIDENCE"

File: supabase/functions/api/_shared/enums.ts
// Ported from backend/app/core/enums.py. Kept as plain string union types
// (not TS enums) since every DB column stores these as plain text/VARCHAR.

export type UserRole = "CITIZEN" | "ADMIN" | "REPAIR_TEAM";

export type DamageType =
  | "POTHOLE"
  | "CRACKED_ROAD"
  | "FLOODING"
  | "DAMAGED_SIDEWALK"
  | "BROKEN_STREETLIGHT"
  | "OTHER"
  | "UNKNOWN";

export type ComplaintStatus =
  | "PENDING"
  | "AI_ANALYZED"
  | "PRIORITIZED"
  | "ASSIGNED"
  | "IN_PROGRESS"
  | "RESOLVED"
  | "REJECTED"
  | "DUPLICATE";

// Allowed status transitions, enforced wherever a status change happens
// (the create_complaint/admin RPCs), so the lifecycle can't be
// short-circuited the same way ComplaintService guarded it in Python.
export const STATUS_TRANSITIONS: Record<ComplaintStatus, ComplaintStatus[]> = {
  PENDING: ["AI_ANALYZED", "PRIORITIZED", "REJECTED", "DUPLICATE"],
  AI_ANALYZED: ["PRIORITIZED", "REJECTED", "DUPLICATE"],
  PRIORITIZED: ["ASSIGNED", "REJECTED", "DUPLICATE"],
  ASSIGNED: ["IN_PROGRESS", "PRIORITIZED", "REJECTED", "DUPLICATE"],
  IN_PROGRESS: ["RESOLVED", "ASSIGNED", "REJECTED"],
  RESOLVED: ["IN_PROGRESS"],
  REJECTED: ["PENDING", "PRIORITIZED"],
  DUPLICATE: ["PENDING", "PRIORITIZED"],
};

export const CLOSED_STATUSES: ComplaintStatus[] = ["RESOLVED", "REJECTED", "DUPLICATE"];

export type PriorityLevel = "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";

export type TrafficLevel = "LOW" | "MEDIUM" | "HIGH" | "VERY_HIGH";

export type PlaceType = "HOSPITAL" | "SCHOOL" | "BUS_STOP" | "MAJOR_INTERSECTION" | "EMERGENCY_SERVICE";

export type AssignmentStatus = "ASSIGNED" | "IN_PROGRESS" | "COMPLETED" | "VERIFIED" | "CANCELLED";

export type DuplicateStatus = "SUGGESTED" | "CONFIRMED" | "REJECTED";

export type NotificationChannel = "IN_APP" | "EMAIL" | "SMS" | "PUSH";

export type ImageKind = "REPORT" | "REPAIR_EVIDENCE";
